compiler/src/dotty/tools/dotc/core/TypeComparer.scala

```diff
@@ -381,7 +403,25 @@ class TypeComparer(@constructorOnly initctx: Context) extends ConstraintHandling
       case tp1: NamedType =>
         tp1.info match {
           case info1: TypeAlias =>
-            if (recur(info1.alias, tp2)) return true
+            def realiasConstraint() = tp2 match {
+              case tp2: TypeParamRef =>
+                constraint.entry(tp2) match {
+                  case TypeBounds(lo, hi) =>
+                    val aliasLo = tp1 != lo && info1.alias == lo
+                    val aliasHi = tp1 != hi && info1.alias == hi
+                    if aliasLo || aliasHi then
+                      constraint = constraint.updateEntry(tp2, TypeBounds(
+                        if aliasLo then tp1 else lo,
+                        if aliasHi then tp1 else hi))
+                  case tp =>
+                    if tp1 != tp && info1.alias == tp then
+                      constraint = constraint.updateEntry(tp2, tp1)
+                }
+              case _ =>
+            }
+            val res = recur(info1.alias, tp2)
+            if (tp1.symbol.isStatic) realiasConstraint()
```

Comment:
We only try to restore the type alias if the symbol is static because then it is guaranteed that it is accessible. Otherwise it might not be. But I suppose even if the symbol is not static, it might be accessible in the current context, right? Is there a better check?

Reply:
There's isAccessibleFrom but doing accessibility checks in the subtype checker seems like too much coupling to me.

Reply:
Thanks for the hint! Then I guess, `isStatic` might be good approximation.